File: Search/InterpolationSearch.rb
def interpolation_search(sample_space, value)
  # Interpolation search [O(log log n)]
  left = 0
  right = sample_space.length - 1

  while left <= right && (left_value = sample_space[left]) <= value && value <= (right_value = sample_space[right])
    pos = left + (value - left_value) * (right - left) / (right_value - left_value)
    return true if sample_space[pos] == value
    
    if sample_space[pos] < value
      left = pos + 1
    else
      right = pos - 1
    end
  end
  
  return false # cannot find value in the list
end

def interpolation_search_recursive(sample_space, value, left = 0, right = sample_space.length - 1)
  # Interpolation search [O(log log n)]
  return false unless left <= right && (left_value = sample_space[left]) <= value && value <= (right_value = sample_space[right])
  
  pos = left + (value - left_value) * (right - left) / (right_value - left_value)
  return true if sample_space[pos] == value
  return interpolation_search_recursive(sample_space, value, pos + 1, right) if sample_space[pos] < value
  return interpolation_search_recursive(sample_space, value, left, pos - 1) if sample_space[pos] > value
end

File: Search/InterpolationSearch.py
def isearch(sample_space, value):
    # Interpolation search [O(log log n)]
    left, right = 0, len(sample_space) - 1

    left_value, right_value = sample_space[left], sample_space[right]
    while left <= right and left_value <= value and value <= right_value:
        pos = left + (value - left_value) * (right - left) // (right_value - left_value)

        if sample_space[pos] == value:
            return pos
        elif sample_space[pos] < value:
            left = pos + 1
        else:
            right = pos - 1

        left_value, right_value = sample_space[left], sample_space[right]
  
    return False # cannot find value in the list


def risearch(sample_space, value, left=0, right=None, debug=False, num_of_checks=0):
    right = right or len(sample_space) - 1
    # Interpolation search [O(log log n)]
    left_value, right_value = sample_space[left], sample_space[right]
    if left > right or left_value > value or value > right_value:
        return False

    pos = left + (value - left_value) * (right - left) // (right_value - left_value)

    if sample_space[pos] == value:
        return pos
    elif sample_space[pos] < value:
        return risearch(sample_space, value, pos + 1, right, debug, num_of_checks)
    else:
        return risearch(sample_space, value, left, pos - 1, debug, num_of_checks)
